{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3868c6ab-5331-4c6d-b5ec-90b0741aac02",
   "metadata": {},
   "source": [
    "# Pangenomes, dictionaries and sets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77443518-5e88-4bd7-a89c-bb3c1ae884aa",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "We have 3 genomes\n",
    "```\n",
    "Genome1 = [1,7,67,834,17,8]\n",
    "\n",
    "Genome2 = [834,8,17,29,7,1]\n",
    "\n",
    "Genome3 = [67,17,834,1,7]\n",
    "```\n",
    "and we want to work out which genes are in all of them (and we bear in mind that later on we want to know which genes are in at least 90% of them). In real life, when we do this kind of this, a real bacterial genome will have around 3000-5000 genes, so this is a toy example of a real-thing.\n",
    "\n",
    "What did our *Simple idea* code look like again?\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43199797-2e0a-4fa0-b947-c0adef47ab92",
   "metadata": {},
   "outputs": [],
   "source": [
    "Genome1 = [1,7,67,834,17,8]\n",
    "Genome2 = [834,8,17,29,7,1]\n",
    "Genome3 = [67,17,834,1,7]\n",
    "core=[]\n",
    "for g in Genome1:\n",
    "    occurrences_of_g = 1\n",
    "    for h in Genome2:\n",
    "        if (g==h):\n",
    "            occurrences_of_g =occurrences_of_g+1\n",
    "    for j in Genome3:\n",
    "        if (g==j):\n",
    "            occurrences_of_g =occurrences_of_g+1\n",
    "            if (occurrences_of_g==3):\n",
    "                core.append(g)\n",
    "\n",
    "#I didn't put anything in the lecture notes about actually printing out the core genome\n",
    "print(\"The core genome, i.e. the set of genes in all 3 genomes, is:\")\n",
    "for c in core:\n",
    "    print(c)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b71e97f-71a7-4c78-8760-65b2959f9e2a",
   "metadata": {},
   "source": [
    "# Stage 1 (i'm putting this here just so we have \"bookmarks\" to refer to)\n",
    "\n",
    "Now, as we mentioned in the lecture, the above code is fine for 3 small toy genomes, but there are definite limitations.\n",
    "If we had to run it on more genomes, we would need to write more loops, one for each extra genome.\n",
    "\n",
    "Right now it says\n",
    "```\n",
    "for h in Genome2:\n",
    "    ..blah blah\n",
    "for j in Genome3:\n",
    "    ..whatever\n",
    "```\n",
    "We would need to explicitly add new lines for each new genome. This is faster and more reliable than manually looking through the files, but it is definitely no fun. So, we'd like a better solution. So, let's store the above code for future use in a function, and then try the other method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f5a70c2-a077-446e-b303-774bb82e0afd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def simple_function(g1, g2, g3):\n",
    "    \"\"\"\n",
    "    Remember a function is like a toaster, or Deliveroo - you pass in specific inputs,\n",
    "    and you get out specific outputs.\n",
    "    Here we are going to pass in 3 genomes g1,g2,g3 , each a list of genes.\n",
    "    The output will be a list of core genes.\n",
    "    \"\"\"\n",
    "    core=[]\n",
    "    for g in g1:\n",
    "        occurrences_of_g = 1\n",
    "        for h in g2:\n",
    "            if (g==h):\n",
    "                occurrences_of_g =occurrences_of_g+1\n",
    "                for j in g3:\n",
    "                    if (g==j):\n",
    "                        occurrences_of_g =occurrences_of_g+1\n",
    "                        if (occurrences_of_g==3):\n",
    "                            core.append(g)\n",
    "\n",
    "    return(core)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9096ca19-b660-467e-8d69-9b259d1cf746",
   "metadata": {},
   "outputs": [],
   "source": [
    "#you can use the function like this, passing in the genomes as lists:\n",
    "c = simple_function([1,7,67,834,17,8] , [834,8,17,29,7,1], [67,17,834,1,7])\n",
    "\n",
    "\n",
    "#or you can pass them in as variables (which were defined up above)\n",
    "d= simple_function(Genome1, Genome2, Genome3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ae08192-97a2-4dea-950e-52782116ccf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(c)\n",
    "print(d)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bb92b1c-6ecb-4f8a-b8db-60cb088dadd5",
   "metadata": {},
   "source": [
    "## Now what about the second approach we talked about in the lecture, using dictionaries?\n",
    "\n",
    "Let's first practice with dictionaries a bit. Here are 2 examples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d55586f5-a243-4bf6-bfa0-66f234504da7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Let's imagine we are keeping track of how many mutations there are in different genes, while processing some information\n",
    "muts={\"gene1\":7 ,\n",
    "     \"gene2\":8,\n",
    "     \"gene3\":100}\n",
    "\n",
    "#what if we want to remember which codon encodes for each amino acid?\n",
    "#(I have not double checked this table is correct, but you should get the idea)\n",
    "#notice how we can intersperse comments using ## to give us more information.\n",
    "codon_to_amino_acid = {\n",
    "    'TTT': 'F', 'TTC': 'F',  # Phenylalanine\n",
    "    'TTA': 'L', 'TTG': 'L',  # Leucine\n",
    "    'CTT': 'L', 'CTC': 'L', 'CTA': 'L', 'CTG': 'L',  # Leucine\n",
    "    'ATT': 'I', 'ATC': 'I', 'ATA': 'I',  # Isoleucine\n",
    "    'ATG': 'M',  # Methionine (Start codon)\n",
    "    'GTT': 'V', 'GTC': 'V', 'GTA': 'V', 'GTG': 'V',  # Valine\n",
    "    'TCT': 'S', 'TCC': 'S', 'TCA': 'S', 'TCG': 'S',  # Serine\n",
    "    'CCT': 'P', 'CCC': 'P', 'CCA': 'P', 'CCG': 'P',  # Proline\n",
    "    'ACT': 'T', 'ACC': 'T', 'ACA': 'T', 'ACG': 'T',  # Threonine\n",
    "    'GCT': 'A', 'GCC': 'A', 'GCA': 'A', 'GCG': 'A',  # Alanine\n",
    "    'TAT': 'Y', 'TAC': 'Y',  # Tyrosine\n",
    "    'TAA': '*', 'TAG': '*',  # Stop codons\n",
    "    'CAT': 'H', 'CAC': 'H',  # Histidine\n",
    "    'CAA': 'Q', 'CAG': 'Q',  # Glutamine\n",
    "    'AAT': 'N', 'AAC': 'N',  # Asparagine\n",
    "    'AAA': 'K', 'AAG': 'K',  # Lysine\n",
    "    'GAT': 'D', 'GAC': 'D',  # Aspartic Acid\n",
    "    'GAA': 'E', 'GAG': 'E',  # Glutamic Acid\n",
    "    'TGT': 'C', 'TGC': 'C',  # Cysteine\n",
    "    'TGA': '*',  # Stop codon\n",
    "    'TGG': 'W',  # Tryptophan\n",
    "    'CGT': 'R', 'CGC': 'R', 'CGA': 'R', 'CGG': 'R',  # Arginine\n",
    "    'AGT': 'S', 'AGC': 'S',  # Serine\n",
    "    'AGA': 'R', 'AGG': 'R',  # Arginine\n",
    "    'GGT': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G',  # Glycine\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a844def8-e90d-4a4c-8d4d-a30db47cf533",
   "metadata": {},
   "source": [
    "Now it's your turn. Can you make a dictionary to remember which days of the week are weekdays,\n",
    "and which are the weekend."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d6db390-1624-4f06-8ce9-88ccd2d7215c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#put your dictionary here. Remember Python wants quotes around text strings, so \"Monday\", not Monday.\n",
    "#I don't really mind, but you could encode weekend as 1 and weekday as 0. Up to you."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5f9473a-6331-4f91-b20a-003d8806b89f",
   "metadata": {},
   "source": [
    "Now let's try a loop, where you add information to a dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd899962-03f0-454f-a963-da138848e13b",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_for_example=[1,2,3,3,3,15,18,9,23,3,56,556,707979,3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e9d7445-d6b0-4df1-8af3-1c53eac10221",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#Suppose we had to write a loop which iterates through data_for_example and counts how many occurrences of the number\n",
    "# 3 there are. For example in [2,2,3] the answer would be 1. For [33,3] the answer would also be 1, we're\n",
    "#just looking for exactly the number 3, Not numbers with 3 in them like 33 or 123.\n",
    "#We would do this\n",
    "count=0\n",
    "for b in data_for_example:\n",
    "    if b==3:\n",
    "        count=count+1\n",
    "#this would work nicely, if we just wanted to count how many 3's there were\n",
    "#what if I want to know how many of ALL the numbers there are?\n",
    "\n",
    "#HINT: again, go through every element of data_for_example.\n",
    "#      if it is NOT in your dictionary, add it. key=number, value=0\n",
    "#       if it is in your dictionary, increment the count\n",
    "\n",
    "#put your loop here, and run it!\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63e4f517-e430-4df7-a09b-84d82860f00b",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## Answer to the previous is here:\n",
    "# make a dictionary - for each number, it will track how often we've seen it\n",
    "dict={}\n",
    "for b in data_for_example:\n",
    "    if b in dict:\n",
    "        dict[b]=dict[b]+1\n",
    "    else:\n",
    "        dict[b]=1 # this is the first time we have seen it\n",
    "#check results\n",
    "for c in dict.keys():\n",
    "    print(c)\n",
    "    print(\" happened \")\n",
    "    print(dict[c])\n",
    "    print(\" times.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b39f364-dc02-4639-8c92-ec716043e860",
   "metadata": {},
   "source": [
    "# We will just introduce a way to loop through a dictionary and get the key-value pairs\n",
    "\n",
    "This is a small dictionary we mentioned above\n",
    "```\n",
    "muts={\"gene1\":7 ,\n",
    "     \"gene2\":8,\n",
    "     \"gene3\":100}\n",
    "````\n",
    "# The items() function, which operates on a dictionary will give you a list of key-value pairs\n",
    "\n",
    "This function is quite useful for dictionaries.\n",
    "```\n",
    "muts.items()\n",
    "```\n",
    " will return\n",
    " \n",
    "```\n",
    " [ (\"gene1\", 7) , (\"gene2\", 8) , (\"gene3\", 100) ]\n",
    "```\n",
    "\n",
    "Apologies for this being weird. Up to now we have talked about functions working like \\\n",
    "```\n",
    "print(\"zam\")\n",
    "```\n",
    "you have the function name, and then the input in brackets. In Python, for some things, such as list, dictionary, the thing has\n",
    "functions of its own. In that case, the function name and inputs go afterwards. Here we have a dictionary called muts, and we call the items function on it (no inputs) like this:\n",
    "\n",
    "```\n",
    "muts.items()\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52e075ad-702d-4415-9386-541564d8343a",
   "metadata": {},
   "source": [
    "# Stage 2 \n",
    "\n",
    "\n",
    "Now let's try a loop where we add information to a dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f3c8521-8b91-49d6-9ec7-917fbe793c25",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Let's imagine we are observing animals moving in and out of a specific\n",
    "# area. We are later going to release a pheromone to see if the types of animal\n",
    "# that appear changes after the pheromone is released.\n",
    "# But for now we just need a baseline. \n",
    "animals=[\"dog\", \"dog\", \"hyrax\",\"tortoise\", \"antelope\",\"cat\", \"unicorn\", \"megatherium\", \"antelope\",\"antelope\",\"antelope\",\"antelope\",\"antelope\",\"antelope\",\"antelope\",\"antelope\",\"antelope\",\"dog\"]\n",
    "\n",
    "#write a loop that goes through this list, and for each type of animal, keeps a count of how many times it is there.\n",
    "# For each element on the list, check if you've seen the animal before. If yes, update a dictionary entry with a new count.\n",
    "#if no, create a new entry in the dictionary\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d78f92d1-a11f-482d-acfc-fdff2669a9f6",
   "metadata": {},
   "source": [
    "# Introducing the final data structure: the set\n",
    "\n",
    "The final step is to take a look at, and run, the second way of calculating the core genome, using dictionaries.\n",
    "We will, for each gene keep track of the genomes which contain it.\n",
    "\n",
    "We could use a LIST, as we have done before. If we did, then\n",
    "```\n",
    "[genome1, genome2]\n",
    "```\n",
    "would be different to\n",
    "```\n",
    "[genome2, genome1]\n",
    "```\n",
    "\n",
    "because order matters in lists. But we just want to know which genomes contain the gene, and don't really care about the order.\n",
    "In Python (and maths) we use a SET for this.\n",
    "```\n",
    "{genome1, genome2}\n",
    "```\n",
    "is a SET of genes, and is the same set as\n",
    "\n",
    "```\n",
    "{genome2, genome1}\n",
    "```\n",
    "\n",
    "We'll use this to make our life easier.\n",
    "\n",
    "## Look at the example in the next cell - it shows you how you can iterate through a dictionary\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85460c03-4967-4784-81e3-57c62d90272e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# In a list, we do something like this to go through them\n",
    "my_list=[\"a\",\"b\",\"c\"]\n",
    "for v in my_list:\n",
    "   print(v)\n",
    "\n",
    "#In a dictionary, we can ALSO go through them all, but a dictionary is a set of PAIRs (key and value)\n",
    "my_dict={ \"a\":1 , \"b\":2, \"c\": 3}\n",
    "for letter, number in my_dict.items():\n",
    "    print(letter)\n",
    "    print(\"--->\")\n",
    "    print(number)\n",
    "    print(\"\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4728dd9a-c001-44b1-a56f-691b7ab37c5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_names={\"g1\":[1,7,67,834,17,8],\n",
    "              \"g2\":[834,8,17,29,7,1],\n",
    "              \"g3\":[67,17,834,1,7]\n",
    "             }\n",
    "#set up an empty dictionary to keep track of which genomes contain each gene\n",
    "genomes_containing={}\n",
    "\n",
    "#I want to iterate through all the keys in the dictionary, and for each key\n",
    "#keep track of what its value is. The items() function returns a list of key-value pairs\n",
    "for genome_name, genome in genome_names.items(): #eg \"g1\" and [1,7,67,834,17,8]\n",
    "    #loop through all the genes in this genome\n",
    "    for gene in genome:\n",
    "        #do a check to see if this gene is already a key in the dictionary\n",
    "        if gene in genomes_containing:\n",
    "            #then we have already got a set of genome names containing this gene. Add our\n",
    "            #current genome_name to the set\n",
    "            genomes_containing[gene].add(genome_name)\n",
    "        else:\n",
    "            #create a new dictionary entry. KEY=this gene, and VALUE=a set containing just this genome\n",
    "            genomes_containing[gene]={genome_name}\n",
    "\n",
    "#Find genes present in all genomes\n",
    "all_genomes=set(genome_names.keys()) #set of all genome names\n",
    "#create (\"initialise\") an empty list\n",
    "genes_in_all_genomes=[] \n",
    "\n",
    "#loop through the dictionary we made above\n",
    "for gene, genome_set in genomes_containing.items():\n",
    "    #check if the gene is in all genomes.\n",
    "    #On the line below, on the left we have genomes this gene is in; on the right we have all genomes\n",
    "    if genome_set==all_genomes:  \n",
    "        genes_in_all_genomes.append(gene) #add the gene to the result list\n",
    "\n",
    "#print the results\n",
    "print(\"Genes present in all genomes\", genes_in_all_genomes)\n",
    "        \n",
    "            \n",
    "            \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d430fb5-b5f5-4117-a589-9ac979e96689",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(genome_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d8e817e-d3f5-49aa-b4cf-7ffa2fbd096f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
