{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "360d736e-a990-4a26-99df-adcfa3d0d8e7",
   "metadata": {},
   "source": [
    "\n",
    "# Parsing a VCF file of mutations \n",
    "\n",
    "\n",
    "Let's explore how we can use Python to find things of interest in a file of identified mutations.\n",
    "First, let's look at how you read through a file one line at a time.\n",
    "\n",
    "**The next cell shows a template for doing this - don't actually run it**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05fede64-ac52-435c-a52d-a0fa4ae51e1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "### don't run this, this is just to show you the code.\n",
    "\n",
    "# \"Open\" the file first, and get a \"handle\", which we call f (but you could call anything you like)\n",
    "# we can use f to refer to that file. The \"r\" means Read-Only - we just want to read the file, not write to it.\n",
    "with open(\"your_file.txt\", \"r\") as f:\n",
    "    continue_loop = True  # initial condition\n",
    "\n",
    "    while continue_loop:\n",
    "        line = f.readline() #<<<<<< this means get the next line of the file and stick it in variable called \"line\"\n",
    "        if not line:\n",
    "            # End of file reached, so there is nothing in line\n",
    "            break\n",
    "\n",
    "        # --- whatever our code logic is goes here ---\n",
    "        # Example:\n",
    "        # if \"stop\" in line:\n",
    "        #     continue_loop = False\n",
    "        # or \"if this is a non-synonymous mutation then do blah\"\n",
    "        # ----------------------------\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13b3a88b-6577-4fde-96eb-270ad3603e19",
   "metadata": {},
   "source": [
    "Right, so let's use this idiom to read the VCF file that I have put in your directory.\n",
    "First thing's first - go to a terminal and look at the file using \"less\". Talk to Zam \n",
    "if you don't know how to do this.\n",
    "\n",
    "\n",
    "Now - this is how to read through the file and print every line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d5e9f9c-9cf2-4420-b0f6-98af3f16c5ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open the file before the loop\n",
    "with open(\"test.vcf\", \"r\") as f:\n",
    "    continue_loop = True  # initial condition\n",
    "\n",
    "    while continue_loop:\n",
    "        line = f.readline()\n",
    "        if not line:\n",
    "            # End of file reached\n",
    "            break\n",
    "        #let's double check it works, and print out each line of the VCF\n",
    "        print(line.strip())\n",
    "\n",
    "\n",
    "##when you run this cell, you will see a VCF file printed out:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fdb6694-85d4-45b8-9746-64b158cb1181",
   "metadata": {},
   "source": [
    "See what happens if you delete the \".strip()\" and re-run the cell - what changes?\n",
    "You can check what strip() does here:\n",
    "https://www.w3schools.com/python/ref_string_strip.asp\n",
    "See if you can understand what is happening"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78e48e62-5358-42a1-8454-8bede4cc7bb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# OK - you have now seen that we can go through\n",
    "# all the lines in the VCF file.\n",
    "# We are going to go through these one at a time, looking at\n",
    "# the position (column 2) and if it lies in the gene, look\n",
    "# for PSTOP in column 10.\n",
    "\n",
    "\n",
    "#define the start and end coordinates of our gene\n",
    "start=150\n",
    "end=201\n",
    "\n",
    "\n",
    "# Open the file before the loop\n",
    "with open(\"test.vcf\", \"r\") as f:\n",
    "    continue_loop = True  # initial condition\n",
    "    \n",
    "    \n",
    "    # Before we iterate through the list, there is a fiddly thing to resolve.\n",
    "    # We don't care about the first line of the VCF\n",
    "    # because it just says CHROM POS ID, etc - it just gives column headers.\n",
    "    # So let's get the line, and then ignore it - we don't even store it in a variable\n",
    "    f.readline()\n",
    "    \n",
    "    #now we do the same thing to all of the remaining lines of the file\n",
    "    while continue_loop:\n",
    "        line = f.readline()\n",
    "        if not line:\n",
    "            # End of file reached\n",
    "            break\n",
    "\n",
    "        #Now I want to split the line into an array of entries (\"chr\", \"100\", etc)\n",
    "        entries = line.split(\"\\t\")\n",
    "       #now I have split out the columns of this line into a list called entries\n",
    "       #I used the  \"split\" command, because I knew the VCF file columns are separated by\n",
    "       # \"tab\" characters, which Python codes as \\t\n",
    "\n",
    "    \n",
    "    # The SECOND issue to tidy, is this\n",
    "    # I said Python allows lists to contain numbers (integers or decimals)\n",
    "    # or text (which is called \"strings\" in computer science).\n",
    "    # when we split the long string into a list of lines, Python got the message\n",
    "    # that all the entries in the list were strings, not numbers\n",
    "    # This means, that when we look at column 2, Python thinks it is text. \n",
    "    # so it won't let us do greater than or less than on it. So we have to \"force\"\n",
    "    # Python to treat it as a number using the \"int\" function that turns it to an integer:\n",
    "        pos = int(entries[1])\n",
    "\n",
    "        #get the tenth column:\n",
    "        tenthcol=entries[9].strip()\n",
    "        if (pos>start) and (pos<=end):\n",
    "            #then we are in the gene\n",
    "            if (tenthcol==\"PSTOP\"):\n",
    "                print(\"I found a pstop at this coordinate in the gene\")\n",
    "                print(pos)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ab1fa80-f8e9-488b-be42-f975f1c33eac",
   "metadata": {},
   "source": [
    "# Exercises\n",
    "\n",
    "1. Can you make a list of 10 numbers, and then write a loop to print every other number?\n",
    "2. Can you make a list of 10 names including yours, and write a loop to go through until it hits your name, and print \"Hurray\"\n",
    "3. Can you write some code to read through the VCF file and print any line with an A->T mutation (ie reference allele is A, and the alternate allele is T)\n",
    "4. Can you find any T->C mutations between coordinates 1000 and 2000 ?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa39ffdd-b578-4530-82ad-70fa467b52d2",
   "metadata": {},
   "source": [
    "# Functions\n",
    "\n",
    "Reminder, this is how you define a function in Python\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5178f158-fb02-4c3c-87a4-7e5218595913",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_minimum_of_list(a_list):\n",
    "    \"\"\"\n",
    "    This function takes one input (CompSci jargon for an input is an ARGUMENT)\n",
    "    argument 1: a list\n",
    "    \"\"\"\n",
    "    min = a_list[0]\n",
    "    for i in a_list:\n",
    "        if i < min:\n",
    "            min = i\n",
    "    return min\n",
    "\n",
    "\n",
    "test = [102, 2, 203]\n",
    "print(get_minimum_of_list(test))\n",
    "\n",
    "\n",
    "# Here is a template you can use\n",
    "def my_function(input1, input2):\n",
    "    \"\"\"\n",
    "    some description here\n",
    "    \"\"\"\n",
    "    # here, do whatever you need to do to define your output\n",
    "    output = 0 #or whatever you want\n",
    "    return output"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e5e367d-99f4-4e73-afe6-aa7595df2630",
   "metadata": {},
   "source": [
    "# Your turn again\n",
    "\n",
    "Can you make a function from your code which looks through the VCF and checks if there is PSTOP in the gene?\n",
    "\n",
    "Have a think about how you might change it so the start and end coordinates could be passed into the function, rather than\n",
    "coding them into the actual function - that way you could use this for different genes. For example, what if you wanted to check for PSTOPs in two different genes, your original one between 150 and 201, and another between 2000 and 3000. Can you write a function to do this?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24de5cd6-feb3-43d6-9639-fb93f276189d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#your code here"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6713038e-64c3-4e04-a2c7-005c19f332fe",
   "metadata": {},
   "source": [
    "Can you write a function that takes a data array and returns histogram data?\n",
    "\n",
    "https://www.codewars.com/kata/5704bf9b38428f1446000a9d\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c4bd3fb-9230-490a-af49-9b2601cf13fe",
   "metadata": {},
   "source": [
    "Something I have not really introduced is how to detect characters/words in strings.\n",
    "eg how to spot \"cat\" in \"catastrophe\".\n",
    "\n",
    "you don't need it below, but fyi, check out python regex:\n",
    "https://www.w3schools.com/python/python_regex.asp\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56435c0d-617a-4310-8c52-76e431fefc33",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Let's just look at working our way through a string. Look what happens when you run this\n",
    "\n",
    "for c in \"abc\":\n",
    "    print(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60508081-8a7b-4a13-8ac3-d77e4484a612",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# This is fine but you donâ€™t know what position (index) each character is at unless you manually track it.\n",
    "\n",
    "#now try this\n",
    "\n",
    "for i, c in enumerate(\"abc\"):\n",
    "    print(i, c)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9ab3750-7934-44a4-b3f3-d3a9a575e874",
   "metadata": {},
   "source": [
    "Hopefully you can see that enumerate allows you to get the individual characters and their positions. You can also apply it to lists. You will need this in the next lecture"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78f67d58-fffd-418a-844e-1d568e163e9b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "mylist=[\"zam\",\"is\",\"best\"]\n",
    "for i,w in enumerate(mylist):\n",
    "    print(i,w)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f20a194e-6c50-4380-8f5a-10043e5d2937",
   "metadata": {},
   "source": [
    "Great - now why don't you have a go at this\n",
    "\n",
    "https://www.codewars.com/kata/57ee24e17b45eff6d6000164\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a6f5422-b4bc-4365-b838-9e893de0b971",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#A hint for the above. \n",
    "\n",
    "mystring = \"abcadd\"\n",
    "char = \"a\"\n",
    "\n",
    "#which positions is there an 'a' in mystring\n",
    "\n",
    "positions = [i for i, c in enumerate(mystring) if c == char]\n",
    "print(positions)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21edabbf-fe64-42e3-ac8d-d43d9d718e6f",
   "metadata": {},
   "source": [
    "More useful codewars examples that are worth trying to \n",
    "https://www.codewars.com/kata/58dbdccee5ee8fa2f9000058\n",
    "\n",
    "https://www.codewars.com/kata/56a4addbfd4a55694100001f\n",
    "\n",
    "https://www.codewars.com/kata/570597e258b58f6edc00230d\n",
    "\n",
    "https://www.codewars.com/kata/573f5c61e7752709df0005d2\n",
    "\n",
    "https://www.codewars.com/kata/56cd44e1aa4ac7879200010b\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
